# Python Lab 3 - Data Manipulation

## Short notes about assignments

- Solutions to assignments are posted after they are graded. 
- All assignments will be posted to Codio at least by Wednesday midnight. Except for HW2, you will have to use Codio to do your assignments.
- I will stay after class every week to answer questions. If you have any last minute questions about your assignments, please ask then. However, I would recommend you to come to my office hours, as I can't stay for a long time after class.

## Dealing with NaNs

While it is not necessary to deal with NaNs in Problem 3 HW2, it is possible to delete the NaNs before doing the analysis.

You write:

```Python
df.dropna()
```
This deletes all the NaNs in the dataframe. Instead, you can use this code to delete NaNs in a specific column:
```Python
df.dropna(subset=['Column name'])
```

## Example: Obesity Data

Remember from previous lab that we dealt with obesity data. you can write the following if statements to deal with NaNs:

In [2]:
import pandas as pd
df = pd.read_csv("Data/BMI_Data.csv")

In [3]:
def function(row):
    if row['BMI'] >= 40:
        return 'Severe Obesity'
    elif row['BMI'] >= 30:
        return 'Obesity'
    elif row['BMI'] >= 25:
        return 'Overweight'
    elif row['BMI'] >= 18.5:
        return 'Normal'
    elif row['BMI'] < 18.5:
        return 'Underweight'
    else:
        return 'NaN'

In [4]:
df['Obesity'] = df.apply(function, axis=1)

In [5]:
df

Gender  Age  Height  Weight    Race1    BMI      Obesity
0       male   34   164.7    87.4    White  32.22      Obesity
1       male   34   164.7    87.4    White  32.22      Obesity
2       male   34   164.7    87.4    White  32.22      Obesity
3       male    4   105.4    17.0    Other  15.30  Underweight
4     female   49   168.4    86.7    White  30.57      Obesity
...      ...  ...     ...     ...      ...    ...          ...
9995    male   28   177.3    92.3  Mexican  29.40   Overweight
9996  female    0     NaN     6.7    White    NaN          NaN
9997    male   27   175.8    96.7  Mexican  31.30      Obesity
9998    male   60   168.8    78.4    White  27.50   Overweight
9999    male   60   168.8    78.4    White  27.50   Overweight

[10000 rows x 7 columns]

In [6]:
df['Obesity'].value_counts()

Normal            2941
Overweight        2656
Obesity           2281
Underweight       1271
Severe Obesity     485
NaN                366
Name: Obesity, dtype: int64

Realize that the NaNs are dealt now. Instead of this, I can remove the NaNs at the beginning.

In [7]:
# It is recommended to specify a name to the modified dataframe
# as you would like to preserve the original dataframe
df_new = df.dropna(subset=['BMI'])

In [8]:
df_new

Gender  Age  Height  Weight    Race1    BMI      Obesity
0       male   34   164.7    87.4    White  32.22      Obesity
1       male   34   164.7    87.4    White  32.22      Obesity
2       male   34   164.7    87.4    White  32.22      Obesity
3       male    4   105.4    17.0    Other  15.30  Underweight
4     female   49   168.4    86.7    White  30.57      Obesity
...      ...  ...     ...     ...      ...    ...          ...
9994    male   28   177.3    92.3  Mexican  29.40   Overweight
9995    male   28   177.3    92.3  Mexican  29.40   Overweight
9997    male   27   175.8    96.7  Mexican  31.30      Obesity
9998    male   60   168.8    78.4    White  27.50   Overweight
9999    male   60   168.8    78.4    White  27.50   Overweight

[9634 rows x 7 columns]

Observe that there are only 9634 rows (10000-366) in this new dataframe. This code deleted 366 NaNs in the data.

In [9]:
df_new['Obesity'] = df_new.apply(function, axis=1)

C:\Users\daiki\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [10]:
df_new['Obesity'].value_counts()

Normal            2941
Overweight        2656
Obesity           2281
Underweight       1271
Severe Obesity     485
Name: Obesity, dtype: int64

No more NaNs observed.

## One step further: Obtaining frequency table

`.value_count()` can be used to find out the raw counts. We will now be creating a frequency table for this data.

```Python
table = pd.crosstab(index=df['Column Name'], columns='count')
```
Can create a frequency table. Specify the column name and the data frame after `index = ` and don't change `columns='count'`

In [13]:
table = pd.crosstab(index=df_new["Obesity"], columns="count") 

In [14]:
table

col_0           count
Obesity              
Normal           2941
Obesity          2281
Overweight       2656
Severe Obesity    485
Underweight      1271

The obesity levels are categorized in aphabetical order. We can use
```Python
df = df.reindex([#You put the order of rows here])
```

In [15]:
table = table.reindex(['Underweight','Normal','Overweight','Obesity','Severe Obesity'])

In [16]:
table

col_0           count
Obesity              
Underweight      1271
Normal           2941
Overweight       2656
Obesity          2281
Severe Obesity    485

## Two way frequency table

Obtaining two way frequency table is very similar as in the previous step. Let's examine obesity broken down by gender.

We use:
```Python
table = pd.crosstab(index=df['Column Name'], columns=df['Column Name'])
```
Instead of `count`, we specify the column

In [26]:
table = pd.crosstab(index=df_new["Obesity"], columns=df["Gender"]) 

In [27]:
table

Gender          female  male
Obesity                     
Normal            1636  1305
Obesity           1100  1181
Overweight        1169  1487
Severe Obesity     311   174
Underweight        625   646

In [28]:
table = table.reindex(['Underweight','Normal','Overweight','Obesity','Severe Obesity'])

In [29]:
table

Gender          female  male
Obesity                     
Underweight        625   646
Normal            1636  1305
Overweight        1169  1487
Obesity           1100  1181
Severe Obesity     311   174

You can also make the index to be a column in dataframe

In [31]:
table['Obesity'] = table.index

In [32]:
table

Gender          female  male         Obesity
Obesity                                     
Underweight        625   646     Underweight
Normal            1636  1305          Normal
Overweight        1169  1487      Overweight
Obesity           1100  1181         Obesity
Severe Obesity     311   174  Severe Obesity